scmgr/app.go

```diff
@@ -149,11 +67,125 @@ func main() {
 		log.SetDebugVisible(c.Int("debug"))
 		return nil
 	}
-	cliApp.Run(os.Args)
+	log.ErrFatal(cliApp.Run(os.Args))
 }
 
+// adminLink tries to store our public key in the conode
+func adminLink(c *cli.Context) error {
+	if pin := c.String("pin"); pin != "" {
+		return errors.New("pin entry not implemented yet")
+	}
+	private := c.String("private")
+	if private == "" {
+		return errors.New("give either pin or private.toml of conode")
+	}
+	var remote struct {
+		Private string
+		Public  string
+		Address network.Address
+	}
+	_, err := toml.DecodeFile(private, &remote)
+	if err != nil {
+		return errors.New("error while reading private.toml: " + err.Error())
+	}
+	pkey, err := encoding.StringHexToScalar(skipchain.Suite, remote.Private)
+	if err != nil {
+		return errors.New("couldn't decode private key: " + err.Error())
+	}
+	pubkey, err := encoding.StringHexToPoint(skipchain.Suite, remote.Public)
+	if err != nil {
+		return errors.New("couldn't decode public key: " + err.Error())
+	}
+	cfg := getConfigOrFail(c)
+	si := network.NewServerIdentity(pubkey, remote.Address)
+	cfg.Link[si.Public.String()] = &link{
+		Private: pkey,
+		Address: remote.Address,
+		Conode:  si,
+	}
+	cerr := skipchain.NewClient().CreateLinkPrivate(si, pkey, skipchain.Suite.Point().Mul(pkey, nil))
+	if cerr != nil {
+		log.Error(cerr)
+		return cerr
+	}
+	log.Lvl1("Correctly linked with", remote.Address)
+	return cfg.save(c)
+}
+
+func adminAuth(c *cli.Context) error {
+	if c.NArg() < 2 {
+		return errors.New("give the (public_hex|IP:Port) and level as argument")
+	}
+	lvl, err := strconv.Atoi(c.Args().Get(1))
+	if err != nil {
+		return err
+	}
+	cfg := getConfigOrFail(c)
+	if len(cfg.Link) == 0 {
+		return errors.New("no link yet")
+	}
+	publicOrIP := c.Args().First()
+	// Check if it's a public key and if we can find it
+	other := cfg.Link[publicOrIP]
```

Comment:
My private.toml file has Public = "big-thing-in-hex" and when I did "scmrg admin auth big-thing-in-hex 1" it worked. But my public.toml file is another obvious place where I could find my public key, and it is a base64-encoded version of the hex decoded big-thing-in-hex. I should be able to give the public key in either format.

Reply:
Perhaps we need to get rid of base64 representation of anything. It's a bit shorter, but it's confusing me because in some places we use hex, and in others base64...

Reply:
So @nikkolasg would like to have hex-only everywhere?